File: 435-non-overlapping-intervals/non-overlapping-intervals.py
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        removals = 0
        prev_e = -1_000_000

        for s, e in sorted(intervals):
            if s < prev_e:
                removals += 1
                prev_e = min(e, prev_e)
            else:
                prev_e = e
        return removals

File: 435-non-overlapping-intervals/non-overlapping-intervals.java
class Solution {
    public int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        int removals = 0;
        int prevEnd = Integer.MIN_VALUE;
        for (int i = 0; i <intervals.length; i++) {
            int nextStart = intervals[i][0];
            int nextEnd = intervals[i][1];

            if (prevEnd <= nextStart) {     // no overlap
                prevEnd = nextEnd;
            } else {
                removals++;
                prevEnd = Math.min(prevEnd, nextEnd);
            }
        }
        return removals;
    }
}